### PR TITLE
Simplify and refactor connection establishment

### PR DESCRIPTION
**Pull request by butonic**
When the ldap error certificate changes connection establishment might silently fail. This PR simplifies the code for establishing ldap connections and adds extended diagnostics logging for errors.

### CODE REVIEW

**Comment by PVince81** [comment]
@butonic is this done and ready for review ? seeems the unit tests are a bit light for the new additions.

please clarify so we can take over

**Comment by butonic** [comment]
Unit tests were in place for connection establishment before. I adjusted them because I got rid of `init()`.

**Comment by PVince81** [comment]
so assuming no tasks left to do

@jvillafanez please review

**Comment by PVince81** [comment]
seems needs more work:
```
1) OCA\User_LDAP\ConnectionTest::testBindFails
Failed asserting that exception of type "\OCA\User_LDAP\Exceptions\BindFailedException" is thrown.
```

I guess we can take over from here

**Comment by PVince81** [comment]
@VicDeo please take over and finish, I think there's only a unit test to fix

**Comment by VicDeo** [comment]
well, the following steps shows me that logging becomes worse with this PR:
1. Setup ldap server and import users with `php occ user:sync "OCA\User_LDAP\User_Proxy" -m remove`
2. update ldap host in database to be invalid.
3. try to get any page with basic auth
4. Check logs
### master
```
{"reqId":"Mi8bvk2af7aThtnsudkV","level":1,"time":"2019-02-20T14:30:20+00:00","remoteAddr":"127.0.0.1","user":"--","app":"admin_audit","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"User abagail_quigley attempted to log into ownCloud from IP address: 127.0.0.1 [CLIENT_IP: 127.0.0.1] [USER_AGENT: curl\/7.60.0]"}
{"reqId":"Mi8bvk2af7aThtnsudkV","level":3,"time":"2019-02-20T14:30:21+00:00","remoteAddr":"127.0.0.1","user":"--","app":"user_ldap","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Error when searching: Can't contact LDAP server code -1"}
{"reqId":"Mi8bvk2af7aThtnsudkV","level":3,"time":"2019-02-20T14:30:21+00:00","remoteAddr":"127.0.0.1","user":"--","app":"user_ldap","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Attempt for Paging?  "}
{"reqId":"Mi8bvk2af7aThtnsudkV","level":2,"time":"2019-02-20T14:30:21+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Login failed: 'abagail_quigley' (Remote IP: '127.0.0.1')"}
{"reqId":"Mi8bvk2af7aThtnsudkV","level":0,"time":"2019-02-20T14:30:21+00:00","remoteAddr":"127.0.0.1","user":"--","app":"no app in context","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Current user is not logged in"}
```

### this branch
```
{"reqId":"535f328f-2869-4f95-977a-0d4c7d38b81f","level":0,"time":"2019-02-20T14:32:24+00:00","remoteAddr":"127.0.0.1","user":"D","app":"user_ldap","method":"PROPFIND","url":"\/~deo\/oc-tmp\/remote.php\/dav\/files\/D\/","message":"No DN found for <D> on localhostt"}
{"reqId":"ixCBKwYDP9Oe93RtVeyr","level":1,"time":"2019-02-20T14:33:02+00:00","remoteAddr":"127.0.0.1","user":"--","app":"admin_audit","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"User abagail_quigley attempted to log into ownCloud from IP address: 127.0.0.1 [CLIENT_IP: 127.0.0.1] [USER_AGENT: curl\/7.60.0]"}
{"reqId":"ixCBKwYDP9Oe93RtVeyr","level":3,"time":"2019-02-20T14:33:02+00:00","remoteAddr":"127.0.0.1","user":"--","app":"user_ldap","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Exception: {\"Exception\":\"OC\\\\ServerNotAvailableException\",\"Message\":\"Lost connection to LDAP server.\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/LDAP.php(304): OCA\\\\User_LDAP\\\\LDAP->postFunctionCall()\\n#1 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/LDAP.php(43): OCA\\\\User_LDAP\\\\LDAP->invokeLDAPMethod(*** sensitive parameters replaced ***)\\n#2 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Connection.php(552): OCA\\\\User_LDAP\\\\LDAP->bind(*** sensitive parameters replaced ***)\\n#3 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Connection.php(175): OCA\\\\User_LDAP\\\\Connection->establishConnection()\\n#4 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Access.php(1004): OCA\\\\User_LDAP\\\\Connection->getConnectionResource()\\n#5 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Access.php(1194): OCA\\\\User_LDAP\\\\Access->executeSearch('(&(|(objectclas...', Array, Array, NULL, NULL)\\n#6 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Access.php(929): OCA\\\\User_LDAP\\\\Access->search('(&(|(objectclas...', Array, Array, NULL, NULL)\\n#7 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Access.php(824): OCA\\\\User_LDAP\\\\Access->searchUsers('(&(|(objectclas...', Array, NULL, NULL)\\n#8 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Access.php(797): OCA\\\\User_LDAP\\\\Access->fetchListOfUsers('(&(|(objectclas...', Array)\\n#9 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/User\\\/Manager.php(474): OCA\\\\User_LDAP\\\\Access->fetchUsersByLoginName('abagail_quigley', Array)\\n#10 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/User_LDAP.php(139): OCA\\\\User_LDAP\\\\User\\\\Manager->getLDAPUserByLoginName('abagail_quigley')\\n#11 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/User_Proxy.php(82): OCA\\\\User_LDAP\\\\User_LDAP->checkPassword(*** sensitive parameters replaced ***)\\n#12 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/Proxy.php(145): OCA\\\\User_LDAP\\\\User_Proxy->walkBackends('abagail_quigley', 'checkPassword', Array)\\n#13 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/User_Proxy.php(187): OCA\\\\User_LDAP\\\\Proxy->handleRequest('abagail_quigley', 'checkPassword', Array)\\n#14 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/User\\\/Manager.php(252): OCA\\\\User_LDAP\\\\User_Proxy->checkPassword(*** sensitive parameters replaced ***)\\n#15 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/User\\\/Session.php(519): OC\\\\User\\\\Manager->checkPassword(*** sensitive parameters replaced ***)\\n#16 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/User\\\/Session.php(332): OC\\\\User\\\\Session->loginWithPassword(*** sensitive parameters replaced ***)\\n#17 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/User\\\/Session.php(361): OC\\\\User\\\\Session->login(*** sensitive parameters replaced ***)\\n#18 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/User\\\/Session.php(480): OC\\\\User\\\\Session->logClientIn(*** sensitive parameters replaced ***)\\n#19 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/base.php(968): OC\\\\User\\\\Session->tryBasicAuthLogin(Object(OC\\\\AppFramework\\\\Http\\\\Request))\\n#20 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/AppFramework\\\/Middleware\\\/Security\\\/SecurityMiddleware.php(232): OC::handleLogin(Object(OC\\\\AppFramework\\\\Http\\\\Request))\\n#21 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/AppFramework\\\/Middleware\\\/Security\\\/SecurityMiddleware.php(126): OC\\\\AppFramework\\\\Middleware\\\\Security\\\\SecurityMiddleware->isLoggedIn()\\n#22 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/AppFramework\\\/Middleware\\\/MiddlewareDispatcher.php(88): OC\\\\AppFramework\\\\Middleware\\\\Security\\\\SecurityMiddleware->beforeController(Object(OCA\\\\Files\\\\Controller\\\\ViewController), 'index')\\n#23 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/AppFramework\\\/Http\\\/Dispatcher.php(84): OC\\\\AppFramework\\\\Middleware\\\\MiddlewareDispatcher->beforeController(Object(OCA\\\\Files\\\\Controller\\\\ViewController), 'index')\\n#24 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/AppFramework\\\/App.php(100): OC\\\\AppFramework\\\\Http\\\\Dispatcher->dispatch(Object(OCA\\\\Files\\\\Controller\\\\ViewController), 'index')\\n#25 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(47): OC\\\\AppFramework\\\\App::main('ViewController', 'index', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#26 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/private\\\/Route\\\/Router.php(342): OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#27 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/lib\\\/base.php(906): OC\\\\Route\\\\Router->match('\\\/apps\\\/files\\\/')\\n#28 \\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/index.php(54): OC::handleRequest()\\n#29 {main}\",\"File\":\"\\\/home\\\/deo\\\/public_html\\\/oc-tmp\\\/apps\\\/user_ldap\\\/lib\\\/LDAP.php\",\"Line\":335}"}
{"reqId":"ixCBKwYDP9Oe93RtVeyr","level":2,"time":"2019-02-20T14:33:02+00:00","remoteAddr":"127.0.0.1","user":"--","app":"core","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Login failed: 'abagail_quigley' (Remote IP: '127.0.0.1')"}
{"reqId":"ixCBKwYDP9Oe93RtVeyr","level":0,"time":"2019-02-20T14:33:02+00:00","remoteAddr":"127.0.0.1","user":"--","app":"no app in context","method":"HEAD","url":"\/~deo\/oc-tmp\/index.php\/apps\/files\/","message":"Current user is not logged in"}
```

File: lib/Connection.php [hunk 1]
```diff
@@ -482,110 +472,125 @@ private function validateConfiguration() {
...
+					$this->configuration->ldapAgentName,
+					$this->configuration->ldapAgentPassword
+				)) {
+					$this->checkError('Bind success', ['app' => __METHOD__.':'.__LINE__]);
+					return true;
 				}
-			} catch (BindFailedException $e) {
-				if (\trim($this->configuration->ldapBackupHost) === "") {
-					throw $e;
+				$this->checkError('Bind failed', ['app' => __METHOD__.':'.__LINE__]);
```
Comment (VicDeo) [inline lib/Connection.php]: this line is not reachable  when bind is failed. In this case `OC\ServerNotAvailableException` exception is thrown and we   are falling into the catch clause (Line 563 below) where this exception is rethrown if there are no more connections

File: lib/Connection.php [hunk 2]
```diff
@@ -482,110 +472,125 @@ private function validateConfiguration() {
...
+	 */
+	private function checkError($msg = 'LDAP Error check', $context = ['app' => __METHOD__]) {
+		$errno = $this->getLDAP()->errno($this->ldapConnectionRes);
+		if ($errno === self::LDAP_SUCCESS) {
+			\OC::$server->getLogger()->debug("$msg: ($errno), ".\var_export($this->ldapConnectionRes, true), $context);
+			return;
+		}
+		$errmsg = $this->getLDAP()->error($this->ldapConnectionRes);
+		if (\is_resource($this->ldapConnectionRes)) {
+			$this->getLDAP()->getOption($this->ldapConnectionRes, LDAP_OPT_DIAGNOSTIC_MESSAGE, $errdiag);
```
Comment (VicDeo) [inline lib/Connection.php]: won't work. `$errdiag` should be passed by reference, but `func_get_args` is limited to get the values only https://github.com/owncloud/user_ldap/blob/master/lib/LDAP.php#L282

**Comment by VicDeo** [comment]
As far as I understand all diagnostics is currently incapsulated into 
`OCA\User_LDAP::postFunctionCall` https://github.com/owncloud/user_ldap/blob/master/lib/LDAP.php#L312-L337
In addition postFunctionCall has some additional checks (e.g. is there a valid resource) 
So the approach taken here is questionable. I mean there is no need to add new method to the ILDAPWrapper interface, implement it and call in Connection.php.

We can just add whatever we need into postFunctionCall and check if the last action was `bind`
or even put it into the `public function bind` itself

**Comment by butonic** [comment]
@VicDeo agreed. the error checking should be done in the ldap wrapper class and proper exceptions should be thrown. I would not add it in the bind function but keep the error handling in one place. The wrepper is supposed to be a thin layer above the native php ldap functions. It should not contain any business logic, but converting errors into exceptions is desirable. Go for it!

File: lib/ILDAPWrapper.php [hunk 3]
```diff
@@ -27,6 +27,12 @@
 namespace OCA\User_LDAP;
 
 interface ILDAPWrapper {
+	const LDAP_SUCCESS = 0x00;
```
Comment (jvillafanez) [inline lib/ILDAPWrapper.php]: Better include a link to the documentation where these codes come from. I guess they're error codes from the ldap library, but it's difficult to know, plus I don't know if there are more codes that could be used (not now), or if we could use whatever we want.

**Comment by jvillafanez** [comment]
For the backup server logic, I'd recommend to try to create a new class, something like `ConnectionDataFetcher` and move that logic there. The `establishConnection` function still feels heavy, so it would be nice to split the complexity of the function.

File: lib/ILDAPWrapper.php (hunk 3 above)
Comment (VicDeo) [reply lib/ILDAPWrapper.php]: done

File: lib/Access.php [hunk 4]
```diff
@@ -1486,11 +1485,7 @@ public function areCredentialsValid($name, $password) {
 		if (!$testConnection->setConfiguration($credentials)) {
 			return false;
 		}
-		try {
```
Comment (VicDeo) [inline lib/Access.php]: this fail a bunch  of acceptance tests I guess. - WebDav requests are expect 4xx status instead of 5xx

**Comment by PVince81** [comment]
apparently this created a change of behavior, need to clarify if this was expected or not

**Comment by VicDeo** [comment]
@micbar It drives me nuts to debug failing acceptance tests. I'd better close this PR and  fix https://github.com/owncloud/enterprise/issues/3063 in a shorter manner...

**Comment by VicDeo** [comment]
Constants were moved to https://github.com/owncloud/user_ldap/pull/435
Diagnostics has been moved to https://github.com/owncloud/user_ldap/pull/436

**Comment by jnweiger** [comment]
This was added to milestone 0.16.0 -- but I don't think it was done.

**Comment by DeepDiver1975** [comment]
old -> dead -> close

**Pull request closed without merging**